src/number.rs

```diff
@@ -292,90 +389,352 @@ impl<'de> Deserialize<'de> for Number {
-            N::Float(f) => visitor.visit_f64(f),
+        if self.visited {
+            return Ok(None)
         }
+        self.visited = true;
+        seed.deserialize(NumberFieldDeserializer).map(Some)
+    }
+
+    fn next_value_seed<V>(&mut self, seed: V) -> Result<V::Value, Error>
+    where
+        V: de::DeserializeSeed<'de>,
+    {
+        seed.deserialize(self.number.to_owned().take().unwrap().into_deserializer())
+    }
+}
+
+#[cfg(feature = "arbitrary_precision")]
+struct NumberFieldDeserializer;
+
+#[cfg(feature = "arbitrary_precision")]
+impl<'de> Deserializer<'de> for NumberFieldDeserializer {
+    type Error = Error;
+
+    fn deserialize_any<V>(self, visitor: V) -> Result<V::Value, Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        visitor.visit_borrowed_str(SERDE_STRUCT_FIELD_NAME)
     }
 
     forward_to_deserialize_any! {
-        bool i8 i16 i32 i64 u8 u16 u32 u64 f32 f64 char str string bytes
-        byte_buf option unit unit_struct newtype_struct seq tuple
-        tuple_struct map struct enum identifier ignored_any
+        bool u8 u16 u32 u64 i8 i16 i32 i64 f32 f64 char str string seq
+        bytes byte_buf map struct option unit newtype_struct
+        ignored_any unit_struct tuple_struct tuple enum identifier
     }
 }
 
-macro_rules! from_signed {
-    ($($signed_ty:ident)*) => {
-        $(
-            impl From<$signed_ty> for Number {
-                #[inline]
-                fn from(i: $signed_ty) -> Self {
-                    if i < 0 {
-                        Number { n: N::NegInt(i as i64) }
-                    } else {
-                        Number { n: N::PosInt(i as u64) }
-                    }
+impl FromPrimitive for Number {
+    #[inline]
+    fn from_i64(n: i64) -> Option<Self> {
+        let n = {
+            #[cfg(not(feature = "arbitrary_precision"))]
+            {
+                if n < 0 {
+                    N::NegInt(n)
+                } else {
+                    N::PosInt(n as u64)
                 }
             }
-        )*
-    };
+            #[cfg(feature = "arbitrary_precision")]
+            {
+                let mut buf = Vec::new();
+                itoa::write(&mut buf, n).unwrap();
+                String::from_utf8(buf).unwrap()
+            }
+        };
+        Some(Number { n: n })
+    }
+
+    #[inline]
+    fn from_u64(n: u64) -> Option<Self> {
+        let n = {
+            #[cfg(not(feature = "arbitrary_precision"))]
+            { N::PosInt(n) }
+            #[cfg(feature = "arbitrary_precision")]
+            {
+                let mut buf = Vec::new();
+                itoa::write(&mut buf, n).unwrap();
+                String::from_utf8(buf).unwrap()
+            }
+        };
+        Some(Number { n: n })
+    }
+
+    #[inline]
+    fn from_f64(n: f64) -> Option<Self> {
+        Self::from_f64(n)
+    }
 }
 
-macro_rules! from_unsigned {
-    ($($unsigned_ty:ident)*) => {
+impl ToPrimitive for Number {
```

Comment:
Is the ToPrimitive impl needed by any other code in the PR or is it here for symmetry with FromPrimitive?

Reply:
Mainly for symmetry, and because I think it would be nice for consumers. I really don't see the problem with implementing `FromPrimitive` and `ToPrimitive`... people who use `num-traits` will appreciate it, in fact.

Reply:
People who use num-traits 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, or 0.9 may not appreciate it as much as you think. Look at how much people love the Serde 0.8 impls in the `url` crate.

Reply:
Okay, fair point. But we'll still have the `num-traits` dependency, as before. Its usage just won't be exposed. Are you okay with that?